### PR TITLE
contracts: Add `seal_rent_status`

### PR DESCRIPTION
**Pull request by athei**
#8231 added `seal_rent_params` which exposes all the rent calculation parameters needed in order to implement custom financing models. However, this would require duplicating the rent calculation in contracts. This PR adds an easier way to implement simpler financing models by adding the function `seal_rent_status` which returns:
https://github.com/paritytech/substrate/blob/360b4675d47f7610e9bff51731f49dc481007f5d/frame/contracts/src/rent.rs#L49-L64

This allows to easily check whether a contract is exempt from rent payments by making sure that at the end of contract execution `*_rent == 0`.

## Unstable Interface

Driven by the desire to have an iterative approach in developing new contract interfaces we introduce the concept of an **unstable interface**. Akin to the rust nightly it allows us to add new interfaces but mark them as unstable so that contract languages can experiment with them and give feedback before we stabilize. We need to support interfaces forever and therefore need to do some testing and iterations before we lock ourselves in to one specific interface.

We introduce a new `unstable-interface` feature from the `pallet-contracts` crate which makes those features available. It needs to be forwarded to the crate by the runtime. In this PR we add it to the substrate runtime as `contracts-unstable-interface`. Never run a live chain with this feature enabled.

##  `seal_debug_message`

The recently introduced `seal_debug_message` is made unstable by this PR. It was never released and should go through some testing before we stabilize it. https://github.com/paritytech/ink/pull/792 needs to change its import module once this is in.

### CODE REVIEW

File: frame/contracts/src/wasm/mod.rs [hunk 1]
```diff
@@ -1858,6 +1862,48 @@ mod tests {
...
+			(i32.load (i32.const 0))	;; buffer size
+		)
+	)
+
+	(func (export "deploy"))
+)
+"#;
+
+	#[test]
+	fn rent_status_work() {
```
Comment (cmichi) [inline frame/contracts/src/wasm/mod.rs:1897]:
```suggestion
	fn rent_status_works() {
```

To be consistent with the other `…_works` test.

**Review by cmichi — APPROVED** [review]
Thank you for implementing this!

File: frame/contracts/src/rent.rs [hunk 2]
```diff
@@ -35,56 +35,49 @@ use sp_runtime::{
...
+#[derive(codec::Encode)]
+#[cfg_attr(test, derive(Debug, PartialEq))]
+pub struct RentStatus<T: Config> {
+	/// Required deposit assuming that this contract is the only user of its code.
+	pub max_deposit: BalanceOf<T>,
+	/// Required deposit assuming the code's current refcount.
+	pub current_deposit: BalanceOf<T>,
+	/// Required deposit assuming the specified refcount (None if 0 is supplied).
+	pub custom_refcount_deposit: Option<BalanceOf<T>>,
+	/// Rent that is payed assuming that the contract is the only user of its code.
```
Comment (ascjones) [inline frame/contracts/src/rent.rs:56]:
```suggestion
	/// Rent that is paid assuming that the contract is the only user of its code.
```

File: frame/contracts/src/rent.rs [hunk 3]
```diff
@@ -35,56 +35,49 @@ use sp_runtime::{
...
+	pub max_rent: BalanceOf<T>,
+	/// Rent that is payed given the code's current refcount.
```
Comment (ascjones) [inline frame/contracts/src/rent.rs:58]:
```suggestion
	/// Rent that is paid given the code's current refcount.
```

File: frame/contracts/src/rent.rs [hunk 4]
```diff
@@ -35,56 +35,49 @@ use sp_runtime::{
...
+	pub current_rent: BalanceOf<T>,
+	/// Rent that is payed assuming the specified refcount (None is 0 is supplied).
```
Comment (ascjones) [inline frame/contracts/src/rent.rs:60]:
```suggestion
	/// Rent that is paid assuming the specified refcount (None is 0 is supplied).
```

File: frame/contracts/src/rent.rs [hunk 5]
```diff
@@ -35,56 +35,49 @@ use sp_runtime::{
...
-	fn withdraw(self, account: &T::AccountId) {
-		if let Ok(imbalance) = T::Currency::withdraw(
-			account,
-			self.amount,
-			WithdrawReasons::FEE,
-			ExistenceRequirement::KeepAlive,
-		) {
-			// This should never fail. However, let's err on the safe side.
-			T::RentPayment::on_unbalanced(imbalance);
+/// We cannot derive `Default` because `T` does not necessarily implement `Default`.
```
Comment (ascjones) [inline frame/contracts/src/rent.rs:66]: Would `frame_support::DefaultNoBound` do the job?

File: frame/contracts/src/rent.rs [hunk 6]
```diff
@@ -532,4 +323,301 @@ where
...
+		free_balance: &BalanceOf<T>,
+		contract: &AliveContractInfo<T>,
+		aggregated_code_size: u32,
+		current_refcount: u32,
+		at_refcount: u32,
+	) -> RentStatus<T> {
+		let calc_share = |refcount: u32| {
+			aggregated_code_size.checked_div(refcount).unwrap_or(0)
+		};
+		let max_share = calc_share(1);
```
Comment (ascjones) [inline frame/contracts/src/rent.rs:338]: Why not just `let max_share = aggregated_code_size;`?

**Review by ascjones — APPROVED** [review]
LGTM

File: frame/contracts/src/rent.rs (hunk 5 above)
Comment (athei) [reply frame/contracts/src/rent.rs]: Nice. Did not know about this. Was able to remove some boilerplate.

File: frame/contracts/src/rent.rs (hunk 6 above)
Comment (athei) [reply frame/contracts/src/rent.rs]: Yes. `max_share` is superflous.

**Comment by athei** [comment]
I will merge soon but I want to do some manual testing whether the forward of the unstable feature works when running the node and then add the feature to the test.

**Comment by athei** [comment]
bot merge

**Pull request merged**